index.bs

```diff
@@ -371,28 +403,164 @@ model]]:
+    <var ignore>navigable</var> and <var ignore>sourceDocument</var>.
+
+<div algorithm>
+
+To <dfn>process navigation start for bounce tracking</dfn> given a [=navigable=]
+|navigable| and [=Document=] |sourceDocument|, perform the following steps:
+
+1. If |navigable| is not a [=top-level traversable=], then abort these steps.
+1. Let |origin| be |sourceDocument|'s [=Document/origin=].
+1. If |origin| is an [=opaque origin=], set |initialHost| to empty string.
+1. Otherwise,
+    1. Let |site| be the result of running [=obtain a site=] given |origin|.
+    1. Set |initialHost| to |site|'s [=host=].
+1. If |navigable|'s [=top-level traversable/bounce tracking record=]'s [=bounce tracking record/client bounce detection running=] is false:
+    1. Set |navigable|'s [=top-level traversable/bounce tracking record=] to a new
+        [=bounce tracking record=] with [=bounce tracking record/initial host=]
+        set to |initialHost|.
+1. Otherwise,
+    1. Let |isUserInitiated| denote whether this navigation was initiated by the user.
+    1. If |isUserInitiated|:
+        1. Run [=record stateful bounces for bounce tracking=].
+        1. Set |navigable|'s [=top-level traversable/bounce tracking record=] to a new [=bounce tracking record=] with
+            [=bounce tracking record/initial host=] set to |initialHost|.
+    1. Otherwise, add |initialHost| to |navigable|'s [=top-level traversable/bounce tracking record=]'s [=bounce tracking record/bounce set=].
+1. Set |navigable|'s [=top-level traversable/bounce tracking record=]'s [=bounce tracking record/client bounce detection running=] to false.
+
+Issue: TODO: Use a concrete spec signal to determine |isUserInitiated|.
+
+</div>
+
+<h5 id="bounce-tracking-mitigations-nav-end-monkey-patch">End Navigation Monkey
+Patch</h5>
+
+At the end of a navigation, fill the [=bounce tracking record/bounce set=].
+
+Insert the following steps in the <a spec="html">navigate</a> algorithm, under "Create navigation params by fetching", after Step 19.7,
+"Wait until either response is non-null...", or as soon as [=response/url list=] becomes available.
+
+1. Run [=process navigation end for bounce tracking=] given
+    <var ignore>navigable</var> and <var ignore>URLs</var>.
+
+<div algorithm>
+
+To <dfn>process navigation end for bounce tracking</dfn> given a [=navigable=]
+|navigable| and a [=response/url list=] |URLs|, perform the following steps:
+
+1. [=Assert=] that |navigable|'s [=top-level traversable/bounce tracking record=]'s
+    [=bounce tracking record/client bounce detection running=] is false.
+1. Run [=run steps after a timeout=] given <var ignore>milliseconds</var> [=client bounce detection timer period=]
+    and <var ignore>completionSteps</var> [=record stateful bounces for bounce tracking=] given |navigable|.
+1. If |navigable| is not a [=top-level traversable=], then abort these steps.
+1. If |URLs| is empty, then abort these steps.
```

Comment:
nit: Do we need to abort or just let the for-each have zero iterations?  Just in the interest of brevity.